### PR TITLE
Interim invoice - Cucumber

### PR DESCRIPTION
**Pull request by fp-ps**

### CODE REVIEW

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/invgroup/interceptor/ModCntrInvoicingGroupRepository.java [hunk 1]
```diff
@@ -86,8 +86,8 @@ public Optional<ProductId> getInvoicingGroupProductFor(@NonNull final ProductId
 				.addOnlyActiveRecordsFilter()
 				.addEqualsFilter(I_ModCntr_InvoicingGroup_Product.COLUMNNAME_M_Product_ID, productId)
 				.andCollect(I_ModCntr_InvoicingGroup_Product.COLUMN_ModCntr_InvoicingGroup_ID)
-				.addCompareFilter(I_ModCntr_InvoicingGroup.COLUMNNAME_ValidFrom, Operator.LESS, Instant.now())
-				.addCompareFilter(I_ModCntr_InvoicingGroup.COLUMNNAME_ValidTo, Operator.GREATER, Instant.now())
+				.addCompareFilter(I_ModCntr_InvoicingGroup.COLUMNNAME_ValidFrom, Operator.LESS, SystemTime.asInstant())
```
Comment (mrie04) [inline backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/invgroup/interceptor/ModCntrInvoicingGroupRepository.java:89]: I think this should be lessOrEqual, because of `de.metas.contracts.modular.invgroup.interceptor.ModCntrInvoicingGroupRepository.validateInvoicingGroupProductNoOverlap`. With `addIntervalIntersection` an invoicingGroupProduct should be selected with validFrom included and validTo excluded.
Also this corner cases should be covered by cucumber.

File: backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/interiminvoice/contract/C_Flatrate_Term_CreateInterimContract_StepDef.java [hunk 2]
```diff
@@ -50,8 +50,8 @@ public C_Flatrate_Term_CreateInterimContract_StepDef(
 		this.flatrateTermTable = flatrateTermTable;
 	}
 
-	@And("create interim contract for modular contract")
-	public void create_interim_contract(@NonNull final DataTable dataTable)
+	@And("invoke \"C_BPartner_InterimContract_Create\" process:")
+	public void invokeC_BPartner_InterimContract_Create_process(@NonNull final DataTable dataTable)
```
Comment (mrie04) [inline backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/interiminvoice/contract/C_Flatrate_Term_CreateInterimContract_StepDef.java:54]: I think this is not correct for `S0306_200` , see `de.metas.contracts.modular.interceptor.C_Flatrate_Term.createInterimContractIfNeeded` vs `de.metas.contracts.modular.interim.invoice.process.C_BPartner_InterimContract_Create`

**Review by mrie04 — CHANGES_REQUESTED** [review]
pls check my comments

File: backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/invgroup/interceptor/ModCntrInvoicingGroupRepository.java (hunk 1 above)
Comment (fp-ps) [reply backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/invgroup/interceptor/ModCntrInvoicingGroupRepository.java:89]: adjusting cucumber scenarios 👍

File: backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/contract/ModCntr_InvoicingGroup_Product_StepDef.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+				}
+			}
+
+			assertThat(errorThrown).isTrue();
+		}
+	}
+
+	private void createModCntrInvoicingGroupProduct(@NonNull final Map<String, String> row)
+	{
+		final I_ModCntr_InvoicingGroup_Product modCntrInvoicingGroupProductRecord = CoalesceUtil.coalesceSuppliers(
```
Comment (adi-stefan) [inline backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/contract/ModCntr_InvoicingGroup_Product_StepDef.java:113]:
```suggestion
		final I_ModCntr_InvoicingGroup_Product modCntrInvoicingGroupProductRecord = CoalesceUtil.coalesceSuppliersNotNull(
```

At this point we're sure this is not null

File: backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/interiminvoice/invoice/C_BPartner_InterimContract_GenerateInterimInvoice_StepDef.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+			final I_C_BPartner_InterimContract bPartnerInterimContractRecord = partnerInterimTable.get(partnerInterimIdentifier);
+
+			final BPartnerInterimContract bPartnerInterimContract = bPartnerInterimContractService.getById(BPartnerInterimContractId.ofRepoId(bPartnerInterimContractRecord.getC_BPartner_InterimContract_ID()));
+
+			final Set<InvoiceCandidateId> generatedICIds = interimInvoiceCandidateService.createInterimInvoiceCandidatesFor(flatrateTermRecord, bPartnerInterimContract);
+			handleGeneratedICs(generatedICIds, row);
+		}
+	}
+
+	private void handleGeneratedICs(@NonNull final Set<InvoiceCandidateId> invoiceCandidateIds, @NonNull final Map<String, String> row)
```
Comment (adi-stefan) [inline backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/interiminvoice/invoice/C_BPartner_InterimContract_GenerateInterimInvoice_StepDef.java:90]: Not so much a handle, but a associate.

File: backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/contract/ModCntr_InvoicingGroup_StepDef.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	private void createModCntrInvoicingGroup(@NonNull final Map<String, String> row)
+	{
+		final String productIdentifier = DataTableUtil.extractStringForColumnName(row, I_ModCntr_InvoicingGroup.COLUMNNAME_Group_Product_ID + "." + TABLECOLUMN_IDENTIFIER);
+		final I_M_Product productRecord = productTable.get(productIdentifier);
+
+		final String name = DataTableUtil.extractStringForColumnName(row, I_ModCntr_InvoicingGroup.COLUMNNAME_Name);
+		final Timestamp validFrom = DataTableUtil.extractDateTimestampForColumnName(row, I_ModCntr_InvoicingGroup.COLUMNNAME_ValidFrom);
+		final Timestamp validTo = DataTableUtil.extractDateTimestampForColumnName(row, I_ModCntr_InvoicingGroup.COLUMNNAME_ValidTo);
+
+		final I_ModCntr_InvoicingGroup modCntrInvoicingGroupRecord = CoalesceUtil.coalesceSuppliers(
```
Comment (adi-stefan) [inline backend/de.metas.cucumber/src/test/java/de/metas/cucumber/stepdefs/contract/ModCntr_InvoicingGroup_StepDef.java:117]:
```suggestion
		final I_ModCntr_InvoicingGroup modCntrInvoicingGroupRecord = CoalesceUtil.coalesceSuppliersNotNull(
```
